src/v/cluster/rm_stm.cc

```diff
@@ -1495,7 +1495,7 @@ ss::future<bool> rm_stm::sync(model::timeout_clock::duration timeout) {
     auto ready = co_await persisted_stm::sync(timeout);
     if (ready) {
         if (_mem_state.term != _insync_term) {
-            _mem_state = mem_state{.term = _insync_term};
+            _mem_state.term = _insync_term;
```

Comment:
no-no-no

imagine leadership bounced back and forth and we updated the term but haven't reset all the other fields, we must reset them

when we replay the log we should cache the `send_tx_leadership` info and during the sync if there is unbalanced `send_tx_leadership` we should first write `receive_tx_leadership` and then update _mem_state. the default mem_state logic should not change

Reply:
This was just a hack to get it working, polished most of this, have a look at the next rev.